File: android/app/src/main/java/com/brekeke/phonedev/utils/PN.java
package com.brekeke.phonedev.utils;

import android.text.TextUtils;
import java.util.Map;

// utils to parse push notification data
// see the related part in rn js for reference
public class PN {
  public static String id(Map<String, String> m) {
    return get(m, "pn-id", "pnId");
  }

  public static String username(Map<String, String> m) {
    return get(m, "to", "pbxUsername");
  }

  public static String tenant(Map<String, String> m) {
    return get(m, "tenant", "pbxTenant");
  }

  public static String host(Map<String, String> m) {
    return get(m, "host", "pbxHostname");
  }

  public static String port(Map<String, String> m) {
    return get(m, "pbxPort", "port");
  }

  public static String displayName(Map<String, String> m) {
    return get(m, "displayname", "displayName");
  }

  public static String from(Map<String, String> m) {
    return get(m, "from");
  }

  public static String image(Map<String, String> m) {
    return get(m, "image");
  }

  public static String imageSize(Map<String, String> m) {
    return get(m, "image_size", "imageSize");
  }

  public static String autoAnswer(Map<String, String> m) {
    return get(m, "autoanswer", "autoAnswer");
  }

  public static String ringtone(Map<String, String> m) {
    return get(m, "ringtone");
  }

  private static String get(Map<String, String> m, String k) {
    // prioritize prefix: x_
    var v = m.get("x_" + k);
    if (!TextUtils.isEmpty(v)) {
      return v;
    }
    return m.get(k);
  }

  private static String get(Map<String, String> m, String k1, String k2) {
    var v = get(m, k1);
    if (!TextUtils.isEmpty(v)) {
      return v;
    }
    return get(m, k2);
  }

  private static String get(Map<String, String> m, String k1, String k2, String k3) {
    var v = get(m, k1, k2);
    if (!TextUtils.isEmpty(v)) {
      return v;
    }
    return get(m, k3);
  }
}

File: ios/BrekekePhone/Utils/PN.swift
import Foundation

extension String {
  func toModel<T: Decodable>(_ type: T.Type) -> T? {
    guard let data = self.data(using: .utf8) else { return nil }
    return try? JSONDecoder().decode(T.self, from: data)
  }
  
  func md5() -> String {
    let data = Data(self.utf8)
    var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
    data.withUnsafeBytes {
      _ = CC_MD5($0.baseAddress, CC_LONG(data.count), &digest)
    }
    return digest.map { String(format: "%02x", $0) }.joined()
  }
}

class PN {
    static func id(_ m: [AnyHashable: Any]) -> String? {
        return get(m, "pn-id", "pnId")
    }

    static func callerName(_ m: [AnyHashable: Any]) -> String {
        if let v = get(m, "displayname", "displayName"), !v.isEmpty {
            return v
        }
        if let from = get(m, "from"), !from.isEmpty {
            return from
        }
        return L.loading()
    }

    static func ringtone(_ m: [AnyHashable: Any]) -> String? {
        return get(m, "ringtone")
    }

    static func username(_ m: [AnyHashable: Any]) -> String? {
        return get(m, "to", "pbxUsername")
    }

    static func tenant(_ m: [AnyHashable: Any]) -> String? {
        return get(m, "tenant", "pbxTenant")
    }

    static func host(_ m: [AnyHashable: Any]) -> String? {
        return get(m, "host", "pbxHostname")
    }

    static func port(_ m: [AnyHashable: Any]) -> String? {
        return get(m, "pbxPort", "port")
    }

    // MARK: - Private helpers

    private static func get(_ m: [AnyHashable: Any], _ k: String) -> String? {
      var d = m["x_\(k)"] as? String ?? ""
      if(d.isEmpty) {
        return m[k] as? String ?? ""
      }
      return d
    }

    private static func get(_ m: [AnyHashable: Any], _ k1: String, _ k2: String) -> String? {
        return get(m, k1) ?? get(m, k2)
    }

    private static func get(_ m: [AnyHashable: Any], _ k1: String, _ k2: String, _ k3: String) -> String? {
        return get(m, k1, k2) ?? get(m, k3)
    }
}
